src/plugins/factoids/factoidsPlugin.js

```diff
@@ -1,36 +1,170 @@
 
-const factoidPlugin = async (msg) => {
-  try {
-    await fs.readFileSync('/tmp/disable-factoids');
-    return;
-  } catch (e) {}
-  if (msg.from === 'ecmabot') {
-    fs.writeFile('/tmp/disable-factoids', 'x', () => {});
+const readFile = promisify(fs.readFile);
+const writeFile = promisify(fs.writeFile);
+
+async function checkAndSaveDisabled(msg) {
+  if (!msg.channel) {
+    return false;
+  }
+
+  const safeTo = slugify(msg.channel);
+
+  const disableFile = `/tmp/disable-factoids${safeTo}`;
+
+  const disabled = await readFile(disableFile).then(() => true, () => false);
+
+  if (!disabled && msg.from === 'ecmabot') {
+    await writeFile(disableFile, 'delete to enable', () => {});
+  }
+
+  if (disabled) {
+    console.log(`Factoids are disabled in ${msg.channel} due to an ecmabot message.`);
+    console.log(`    Delete ${disableFile} to reset this.`);
   }
-  if (!msg.command) return null;
 
-  let parts = msg.command.command.split('@').map((x) => x.trim());
-  const [key, nick] = parts;
+  return disabled;
+}
 
-  const fact = facts[key];
-  if (!fact) return;
+class Command {
+  type = '';
+
+  arg = null;
+
+  constructor(command) {
+    const learn = command.match(/learn\s+([^=]+[^=\s])\s*=\s*(.*)$/);
+
+    if (learn) {
+      const [, key, value] = learn;
+      this.type = 'learn';
+      this.arg = { key, value };
+      return;
+    }
+
+    const forget = command.match(/forget\s+([^=]+[^=\s])\s*$/);
+    if (forget) {
+      const [, key] = forget;
+      this.type = 'forget';
+      this.arg = { key };
+      return;
+    }
+
+    const publish = command.match(/publish\s+([^=]+[^=\s])\s*$/);
+    if (publish) {
+      const [, key] = publish;
+      this.type = 'publish';
+      this.arg = { key };
+      return;
+    }
+
+    this.type = 'factoid';
+    this.arg = { key: command.trim() };
+  }
 
-  let value = fact.value;
-  if (fact.alias) {
-    const fact2 = facts[fact.alias];
-    if (fact2) {
-      value = fact2.value;
+  match(matchers) {
+    const has = Object.prototype.hasOwnProperty;
+    if (has.call(matchers, this.type)) {
+      return matchers[this.type](this.arg);
+    } else {
+      return matchers.default();
     }
   }
-  if (!value) return;
 
-  if (nick) {
-    msg.respond(`${nick}, ${value}`);
-  } else {
-    msg.respondWithMention(value);
+  log(msg) {
+    const log = this.type === 'factoid' ? msg.vlog : msg.log;
+    log(`type=${inspect(this.type)} arg=${inspect(this.arg)}`);
+    return this;
   }
+}
+
+const factoidPlugin = async (msg) => {
+  if (!msg.command) return null;
+  if (await checkAndSaveDisabled(msg)) {
+    return;
+  }
+
+  const command = text.validate(msg.command.command).trim();
+
+  const STORE = getStore();
+
+  if (STORE.needsLoadFromDisk()) {
+    await STORE.loadFromDisk();
+  }
+
+  const moderators = (msg.selfConfig && msg.selfConfig.moderators) || [];
+  console.log({ self: msg.selfConfig });
+  const isModerator = moderators.includes(msg.from);
```

Comment:
is this the only check done? people can circumvent this by changing their nicks to a mod's nick when the mod is not online

checking if the user is identified can resolve this. on freenode you can send `ACC username` to NickServ to see if they are. most other services / servers use `STATUS username` instead

There isnt a standard for this but using ACC or STATUS has convered every server I've tried it on

Reply:
Yeah, would this work?

- cache mapping msg.from (user sending the message) to account name
- invalidate cache for a user on part, quit, nick change
- use ACC or STATUS (based on config) as needed to populate the cache
- failure of ACC/STATUS causes permission to be denied

Reply:
not sure if the cache will buy you much at the cost of complexity and runtime cost

In [my implementation](https://github.com/kirjavascript/nibblrjr/blob/master/irc/evaluate/access.js) I simply check on every elevated request. The difference is basically unnoticable so I had no desire to try and improve the speed.

Reply:
What happens if the bot misses a part, quit, or nick change?

Reply:
How about this?

- no cache (for now)
- check username on !learn
  - if not logged in, reject the !learn
  - otherwise, store the proposed change with the account name rather than nick
- check the username on !publish
  - if not logged in, reject the publish
  - if not in moderator list, reject the publish
  - otherwise make the change live

Reply:
why reject the learn from an anon user? we don't get that problem with ecmabot, and anyone on freenode could edit the factoids directly in a PM, auth or not.

Reply:
I would like to track the account name in the proposals to have a more reliable history of changes. That requires the user to be logged in.

We could alternatively allow anonymous !learn and do something having an `editorNick`, `editorAccount`, with the latter potentially being `null`.

Reply:
Tracking both nick and account seems important regardless; it seems useful to then allow account to be null?

We can always lock it down further later if it becomes a problem.